File: Part1_강의자료/Ch08_투포인터/문제별코드/2003_수들의합2/2003_binary_search.py
N, M = map(int, input().split())
li = [0] + list(map(int, input().split()))
# import itertools
# acc = list(itertools.accumulate(li))
acc = [0] * (N + 1)
for i in range(1, N + 1):
    acc[i] = acc[i - 1] + li[i]

ans = 0
for i in range(1, N + 1):
    l, r = i, N
    while l <= r:
        m = (l + r) // 2
        current_sum = acc[m] - acc[i - 1]
        if current_sum > M:
            r = m - 1
        elif current_sum < M:
            l = m + 1
        else:
            ans += 1
            break

print(ans)

File: Part1_강의자료/Ch08_투포인터/문제별코드/2003_수들의합2/2003_binary_search.cpp
#include <stdio.h>
#include <vector>
using namespace std;

int main()
{
    int N, M, num;
    scanf("%d %d", &N, &M);

    vector<int> acc(N + 1);
    for (int i = 1; i <= N; i++)
    {
        scanf("%d", &num);
        acc[i] = acc[i - 1] + num;
    }

    int ans = 0;
    for (int i = 1; i <= N; i++)
    {
        int l = i, r = N;
        while (l <= r)
        {
            int m = (l + r) / 2;
            long long sum = acc[m] - acc[i - 1];
            
            if (sum > M)
                r = m - 1;
            else if (sum < M)
                l = m + 1;
            else
            {
                ans++;
                break;
            }
        }
    }

    printf("%d\n", ans);
    return 0;
}
